did-sdk/src/ion/ion.ts

```diff
@@ -0,0 +1,340 @@
+import fetch from 'node-fetch'
+import * as IonSdk from '@decentralized-identity/ion-sdk'
+import IonDocumentModel from '@decentralized-identity/ion-sdk/lib/models/IonDocumentModel'
+import ProofOfWorkSDK from 'ion-pow-sdk'
+import { DIDDocument } from 'did-resolver'
+import {
+  generateKeyPair,
+  KeyType,
+  generateFromSeedOptions,
+} from './keypairs.js'
+import { WritableDidDocAPI, ReadOnlyDidDocAPI } from '../did-documents.js'
+import {
+  Op,
+  OpCreate,
+  OpUpdate,
+  OpRecover,
+  OpDeactivate,
+  OpUpdateParams,
+} from './ops.js'
+
+const RESOLVE_ENDPOINT =
+  'https://beta.discover.did.microsoft.com/1.0/identifiers'
+const CHALLENGE_ENDPOINT =
+  'https://beta.ion.msidentity.com/api/v1.0/proof-of-work-challenge'
+const SOLUTION_ENDPOINT = 'https://beta.ion.msidentity.com/api/v1.0/operations'
+
+export { KeyType, generateFromSeedOptions } from './keypairs.js'
+export type CreateParams = IonDocumentModel
+export type UpdateParams = OpUpdateParams
+export type RecoverParams = IonDocumentModel
+export type KeyParams = generateFromSeedOptions & { keyType: KeyType }
+export * as OPS from './ops.js'
+export interface DidIonSerializedState {
+  shortForm: string
+  longForm: string
+  ops: Op[]
+}
+
+interface ResolveResponse {
+  didDocument: DIDDocument
+  didDocumentMetadata: any
+}
+
+export async function resolve(
+  didUri: string,
+  ionResolveEndpoint = RESOLVE_ENDPOINT,
+): Promise<ReadOnlyDidDocAPI> {
+  const res = await fetch(ionResolveEndpoint + '/' + didUri)
+  if (res.status >= 400) throw new Error('Not Found')
+  const resObj = (await res.json()) as ResolveResponse
+  if (!resObj?.didDocument) throw new Error('Not found')
+  return new ReadOnlyDidDocAPI(resObj.didDocument, resObj.didDocumentMetadata)
+}
+
+export async function create(
```

Comment:
Do we have a way to do the signature from another location? 

(putting aside the key management issues of this for a minute, let's just say that a ledger or TPM can sign valid ION updates), could we generate a keypair on the hardware device, format the DID doc, and then send it over to the hardware device for signing?

Reply:
Generally speaking that should be possible; we'll just need to write more custom code.